File: Core/src/main/java/dev/lynith/core/ui/styles/ComponentStyles.java
package dev.lynith.core.ui.styles;

import dev.lynith.core.ui.components.Component;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Getter @Setter @Accessors(chain = true)
public abstract class ComponentStyles<C extends Component<C, CS>, CS extends ComponentStyles<C, CS>> {

    private final C component;

    public ComponentStyles(C component) {
        this.component = component;
    }

    private final Map<String, Style<?, ?>> styles = new HashMap<>();

    public <S extends Style<S, ?>> CS addStyle(S style) {
        styles.put(style.getName(), style);
        return (CS) this;
    }

    public <S extends Style<S, ?>> S getStyle(String name) {
        return (S) styles.get(name);
    }

    public List<Style<?, ?>> getStyles() {
        return new ArrayList<>(styles.values());
    }

}

File: Core/src/main/kotlin/dev/lynith/core/ui/styles/ComponentStyles.kt
package dev.lynith.core.ui.styles

import dev.lynith.core.Platform
import dev.lynith.core.ui.components.Component
import dev.lynith.core.ui.styles.impl.*
import dev.lynith.core.ui.theme.AbstractTheme

abstract class ComponentStyles<C : Component<C, CS>, CS : ComponentStyles<C, CS>>(
    val theme: AbstractTheme,
    val state: ComponentStyleState = ComponentStyleState.NORMAL
) {

    open var position = Position()
    open var border = Border()
    open var backgroundColor = theme.colorScheme.background
    open var color = theme.colorScheme.foreground
    open var font = FontStyles()
    open var padding = Padding()
    open var boxShadow = Shadow()
    open var borderRadius = CornerRadius()

    // recursive generics are brain damaging
    class BaseStyles<C : Component<C, BaseStyles<C>>> : ComponentStyles<C, BaseStyles<C>>(Platform.themeManager.currentTheme, ComponentStyleState.NORMAL)

    open var hoverStyles: Class<out CS>? = null

    fun configure(func: CS.() -> Unit): CS {
        func.invoke(this as CS)
        return this as CS
    }

    enum class ComponentStyleState {
        NORMAL,
        HOVER,
        ACTIVE,
        DISABLED,
    }
}
